test/functional/rpc_signmessage.py

```diff
@@ -11,32 +11,73 @@ class SignMessagesTest(BitcoinTestFramework):
+        self.extra_args = [["-addresstype=bech32"]]
 
     def skip_test_if_missing_module(self):
         self.skip_if_no_wallet()
 
     def run_test(self):
         message = 'This is just a test message'
 
-        self.log.info('test signing with priv_key')
+        self.log.info('test signing with priv_key with bech32')
+        priv_key = 'cVbpmtsSXkBJhSZakwBSUY7jxhUgXZeT4hCAGyKixccRZwMG4jrf'
+        address = 'bcrt1qf2yvxk355rc6pqgprrn7q4d3lctr22ta3ne08w'
+        signature = self.nodes[0].signmessagewithprivkey(priv_key, message, 'bech32')
+        assert self.nodes[0].verifymessage(address, signature, message)
+        assert_equal("VALID", self.nodes[0].verifymessage(address, signature, message, 1))
+
+        self.log.info('test legacy key signing')
         priv_key = 'cUeKHd5orzT3mz8P9pxyREHfsWtVfgsfDjiZZBcjUBAaGk1BTj7N'
         address = 'mpLQjfK79b7CCV4VMJWEWAj5Mpx8Up5zxB'
         expected_signature = 'INbVnW4e6PeRmsv2Qgu8NuopvrVjkcxob+sX8OcZG0SALhWybUjzMLPdAsXI46YZGb0KQTRii+wWIQzRpG/U+S0='
         signature = self.nodes[0].signmessagewithprivkey(priv_key, message)
         assert_equal(expected_signature, signature)
         assert self.nodes[0].verifymessage(address, signature, message)
+        assert_equal("VALID", self.nodes[0].verifymessage(address, signature, message, 1))
 
         self.log.info('test signing with an address with wallet')
         address = self.nodes[0].getnewaddress()
         signature = self.nodes[0].signmessage(address, message)
         assert self.nodes[0].verifymessage(address, signature, message)
+        assert_equal("VALID", self.nodes[0].verifymessage(address, signature, message, 1))
 
         self.log.info('test verifying with another address should not work')
         other_address = self.nodes[0].getnewaddress()
+        legacy_address = self.nodes[0].getnewaddress(address_type='legacy')
+        other_signature = self.nodes[0].signmessage(other_address, message)
+        legacy_signature = self.nodes[0].signmessage(legacy_address, message)
+        assert not self.nodes[0].verifymessage(other_address, signature, message)
+        assert not self.nodes[0].verifymessage(address, other_signature, message)
+        assert not self.nodes[0].verifymessage(legacy_address, signature, message)
+        assert not self.nodes[0].verifymessage(address, legacy_signature, message)
+        assert_equal("INVALID", self.nodes[0].verifymessage(other_address, signature, message, 1))
+        assert_equal("INVALID", self.nodes[0].verifymessage(address, other_signature, message, 1))
+        assert_equal("INVALID", self.nodes[0].verifymessage(legacy_address, signature, message, 1))
+        # We get an error here, because the verifier believes the proof should be a BIP-322 proof, due to
+        # the address not being P2PKH, but the proof is a legacy proof
+        assert_equal("ERROR", self.nodes[0].verifymessage(address, legacy_signature, message, 1))
```

Comment:
I think if the (attacker-provided) input is indeed invalid, this should `INVALID` instead of `ERROR`. I think it's cleaner to use `ERROR` only for real processing errors when we can't reach a result, e.g., out-of-memory (okay, maybe a bad example here). Here, we have a result: The proof is invalid because a legacy proof for P2PKH is not allowed.

The background is that crypto APIs typically work like this and are easier to get wrong for the user. If someone gives me a 3 byte string and claims it's an ECDSA signature for some message and some key, then it's an invalid signature. That's all I care about if I call a RPC, I don't care (or shouldn't care) why it's invalid and I don't need to handle different errors differently.

Reply:
I think it's useful to distinguish between what is considered a proper but invalid BIP-322 proof vs something that is not a proper BIP-322 proof. An error in this case means the proof could not be deserialized. Saying it's invalid means the user will not even know that the proof wasn't checked at all, which would be useful for debugging the reason. (For example, if someone provides a proof from Electrum (IIRC) which uses the legacy mode for new address types, it should error, not say invalid, because those proofs are not properly formatted according to BIP-322.)

Reply:
Hm, this has two layers: What other code (ours or not ours) is doing with the return value and what the user is doing with the return value. 
 * For other code: if the proof is invalid, you should treat it as such, no matter what the reason is. (Note that the proof is attacker-provided. It's not an "uncommon" event to get a totally malformed proof.) Having different codes that are all invalid opens up the possibility for subtle bugs in handling those e.g., exception thrown for error and return value for `INVALID`, which leads to different behavior in the end.
 * For the user: I think the user should not be concerned with debugging information. Also, the error code will not be essentially here, the user can always show the (invalid) proof.

Reply:
I'm not sure I agree that users should not be concerned with debugging information. Oftentimes people are teaching each other how to do these things. If the other person says "it tells me 'invalid'" because they missed one character when copy-pasting the proof (or something), that just adds unnecessary ambiguity to the situation.

Perhaps 'ERROR' is the wrong word, though. 'MALFORMED' might be better..